node/src/components/linear_chain_sync/operations.rs

```diff
@@ -91,34 +96,96 @@ where
     }
 }
 
+/// Fetches and stores a block header from the network
 async fn fetch_and_store_block_header<REv, I>(
     effect_builder: EffectBuilder<REv>,
     block_hash: BlockHash,
-) -> Result<Box<BlockHeader>, LinearChainSyncError<I>>
+) -> Result<Box<BlockHeader>, FetcherError<BlockHeader, I>>
 where
     REv: From<FetcherRequest<I, BlockHeader>> + From<NetworkInfoRequest<I>> + From<StorageRequest>,
     I: Eq + Debug + Clone + Send + 'static,
 {
     let block_header =
         fetch_retry_forever::<BlockHeader, REv, I>(effect_builder, block_hash).await?;
-    if !effect_builder
+    effect_builder
         .put_block_header_to_storage(block_header.clone())
-        .await
+        .await;
+    Ok(block_header)
+}
+
+fn validate_finality_signatures(
+    block_header: &BlockHeader,
+    trusted_validator_weights: &BTreeMap<PublicKey, U512>,
+    finality_threshold_fraction: Ratio<u64>,
+    block_signatures: &BlockSignatures,
+) -> Result<(), FinalitySignatureError> {
+    // Check the signatures' block hash is the header's block hash
+    let block_hash = block_header.hash();
+    if block_signatures.block_hash != block_hash {
+        return Err(
+            FinalitySignatureError::SignaturesDoNotCorrespondToBlockHeader {
+                block_header: Box::new(block_header.clone()),
+                block_hash: Box::new(block_hash),
+                block_signatures: Box::new(block_signatures.clone()),
+            },
+        );
+    }
+
+    // Cryptographically verify block signatures
+    block_signatures.verify()?;
+
+    // Calculate the weight of the signatures
+    let mut signature_weight: U512 = U512::zero();
+    for (public_key, _) in block_signatures.proofs.iter() {
+        match trusted_validator_weights.get(public_key) {
+            None => {
+                return Err(FinalitySignatureError::BogusValidator {
+                    trusted_validator_weights: trusted_validator_weights.clone(),
+                    block_signatures: Box::new(block_signatures.clone()),
+                    bogus_validator_public_key: Box::new(public_key.clone()),
+                })
+            }
+            Some(validator_weight) => {
+                signature_weight += *validator_weight;
+            }
+        }
+    }
+
+    // Check the finality signatures have sufficient weight
+    let total_weight: U512 = trusted_validator_weights
+        .iter()
+        .map(|(_, weight)| *weight)
+        .sum();
+
+    let lower_bound = finality_threshold_fraction.add(1).div(2);
+    // Verify: signature_weight / total_weight >= lower_bound
+    // Equivalent to the following
+    if signature_weight * U512::from(*lower_bound.denom())
+        < total_weight * U512::from(*lower_bound.numer())
     {
-        Err(LinearChainSyncError::CouldNotStoreBlockHeader { block_header })
-    } else {
-        Ok(block_header)
+        return Err(FinalitySignatureError::InsufficientWeightForFinality {
```

Comment:
This would return an error if this particular `BlockSignatures`, received from _a single peer_ does not have more than the required `lower_bound` signatures (by weight) but I don't think it is correct to expect that every peer we query for a block at height will already have enough signatures to pass this test. For example, if the block has just been finalized and/or signatures haven't been gossiped then that peer won't have enough of them. Maybe it doesn't matter in practice as we expect that nodes finalized and exchange signatures quickly enough.

Reply:
That's true:
* Either we need to say we always expect a full set of signatures when requesting by height. Then we should reply "we don't have it" if someone requests a block that we don't yet have enough signatures for.
* Or we send the block anyway, but then we need to start collecting signatures here, on the receiver end, and put the block in some cache.

I guess the first option is easier?

Reply:
> This will return an error if this particular BlockSignatures instance, received from a single peer does not have more than the required lower_bound signatures (by weight) but I don't think it is correct to expect that every peer we query for a block at height will already have enough signatures to pass this test.

This will keep trying until it finds a peer with sufficient weight.

In the future, if a peer does not have sufficient weight it will report the block/block-header as missing, but that is beyond the scope of this PR.